src/statsmodels/formula.jl

```diff
@@ -246,78 +235,145 @@ end
+    if !contrasts
+        contr
+    elseif !(1 <= base <= n)
+        error("base = $base is not allowed for n = $n")
+    else
+        contr[:, vcat(1 : (base-1), (base+1) : end)]
+    end
 end
 contr_treatment(n::Integer,contrasts::Bool,sparse::Bool) = contr_treatment(n,contrasts,sparse,1)
 contr_treatment(n::Integer,contrasts::Bool) = contr_treatment(n,contrasts,false,1)
 contr_treatment(n::Integer) = contr_treatment(n,true,false,1)
-cols(v::PooledDataVector) = contr_treatment(length(v.pool))[v.refs,:]
-cols(v::DataVector) = convert(Vector{Float64}, v.data)
-cols(v::Vector) = convert(Vector{Float64}, v)
-
-function isfe(ex::Expr)                 # true for fixed-effects terms
-    if ex.head != :call error("Non-call expression encountered") end
-    ex.args[1] != :|
-end
-isfe(a) = true
-
+cols(v::PooledDataVector) = contr_treatment(length(v.pool))[v.refs, :]
+cols(v::DataVector) = asMatrix(convert(Vector{Float64}, v.data))
+cols(v::Vector) = asMatrix(convert(Vector{Float64}, v))
+
+"""
+    expandcols(trm::Vector)
+Create pairwise products of columns from a vector of matrices
+"""
 function expandcols(trm::Vector)
     if length(trm) == 1
-        return convert(Array{Float64}, trm[1])
+        asMatrix(convert(Array{Float64}, trm[1]))
     else
         a = convert(Array{Float64}, trm[1])
-        b = expandcols(trm[2:end])
-        nca = size(a, 2)
-        ncb = size(b, 2)
-        return hcat([a[:, i] .* b[:, j] for i in 1:nca, j in 1:ncb]...)
+        b = expandcols(trm[2 : end])
+        reduce(hcat, [broadcast(*, a, view(b, :, j)) for j in 1 : size(b, 2)])
     end
 end
 
-function nc(trm::Vector)
-    isempty(trm) && return 0
-    n = 1
-    for x in trm
-        n *= size(x, 2)
+"""
+    dropRanefTerms(trms::Terms)
+Expressions of the form `(a|b)` are "random-effects" terms and are not
+incorporated in the ModelMatrix.  This function checks for such terms and,
+if any are present, drops them from the `Terms` object.
+"""
+function dropRanefTerms(trms::Terms)
+    retrms = Bool[Meta.isexpr(t, :call) && t.args[1] == :| for t in trms.terms]
+    if !any(retrms)  # return trms unchanged
+        trms
+    elseif all(retrms) && !trms.response   # return an empty Terms object
+        Terms(Any[],Any[],Array(Bool, (0,0)), Int[], false, trms.intercept)
+    else
+        # the rows of `trms.factors` correspond to `eterms`, the columns to `terms`
+        # After dropping random-effects terms we drop any eterms whose rows are all false
+        ckeep = !retrms                 # columns to retain
+        facs = trms.factors[:, ckeep]
+        rkeep = vec(sum(facs, 2) .> 0)
+        Terms(trms.terms[ckeep], trms.eterms[rkeep], facs[rkeep, :],
+            trms.order[ckeep], trms.response, trms.intercept)
     end
-    n
 end
 
+"""
+    dropResponse(trms::Terms)
+Drop the response term, `trms.eterms[1]` and the first row and column
+of `trms.factors` if `trms.response` is true.
+"""
+dropResponse(trms::Terms) = !trms.response ? trms :
```

Comment:
I find the `!` confusing. I thought it was idiomatically used for things that modified their first argument, but `dropresponse!` doesn't (it returns a new `Terms` instance).